### PR TITLE
[VisBuilder] fixes filters for table visualisation

### LINKED ISSUE #2923: [BUG] Table vis allows fields that can cause the app to crash
**Describe the bug**

When a count field is added to the split rowss dropbox and then clicked to edit the agg type, the whole app crashes

**To Reproduce**
1. Go to VisBuilder
2. Drag and Drop count to "Split Rows"
3. Click on the "filters" field (The dropped field)
4. See app crash

**Expected behavior**
- [ ] This field should not be allowed in the split rows dropbox.
- [ ] The app should gracefully error in such a scenario

**OpenSearch Version**
3.0

**Dashboards Version**
3.0

**Plugins**

Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**
 - Browser and version [e.g. 22]

**Additional context**

### PR DESCRIPTION
**Pull request by ashwin-pc**
### Description
- Fixes the filters aggregation for table vis
- Adds error boundary to prevent other issues from crashing the who app.
- Minor: Fixes a TS lint issue for TableVisExpressionFunctionDefinition which is referenced in the visbuilder table aggs fn but never actually exported.

Screen Shot 2023-01-06 at 4 39 20 PM

 
### Issues Resolved
#2923 
 
### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/vis_builder/public/application/components/data_tab/secondary_panel.tsx [hunk 1]
```diff
@@ -98,49 +103,98 @@ export function SecondaryPanel() {
...
+          })}
+          color="danger"
+          iconType="alert"
+        >
+          <p>
+            {i18n.translate('visBuilder.aggParamsEditor.errorMsg', {
+              defaultMessage: 'Something went wrong while editing the aggregation',
+            })}
+          </p>
+        </EuiCallOut>
```
Comment (joshuarrrr) [inline src/plugins/vis_builder/public/application/components/data_tab/secondary_panel.tsx:195]: Does this need UX signoff? Why not use a toast as we are for other errors?

File: src/plugins/vis_builder/public/types.ts [hunk 2]
```diff
@@ -37,6 +37,7 @@ export interface VisBuilderPluginStartDependencies {
 }
 
 export interface VisBuilderServices extends CoreStart {
+  appName: string;
```
Comment (joshuarrrr) [inline src/plugins/vis_builder/public/types.ts:40]: The changes to set `appName` in VisBuilderServices seems fine, but it's unclear why it's necessary for this PR. I didn't see it used in `secondary_panel.tsx`?

File: src/plugins/vis_type_table/public/index.ts [hunk 3]
```diff
@@ -9,3 +9,6 @@ import { TableVisPlugin as Plugin } from './plugin';
 export function plugin(initializerContext: PluginInitializerContext) {
   return new Plugin(initializerContext);
 }
+
+/* Public Types */
+export { TableVisExpressionFunctionDefinition } from './table_vis_fn';
```
Comment (joshuarrrr) [inline src/plugins/vis_type_table/public/index.ts:14]: Why? Appears unused in the rest of this PR.

**Review by joshuarrrr — COMMENTED** [review]
What about tests? It also seems like there are some unrelated changes here.

Having an error boundary that prevents the app from crashing is definitely good. But how can we ensure that the visualization type definitions don't lead to these error conditions? It's still not particularly helpful to the user if the UI lets them try to do something that will never work.

File: src/plugins/vis_builder/public/application/components/data_tab/secondary_panel.tsx (hunk 1 above)
Comment (ashwin-pc) [reply src/plugins/vis_builder/public/application/components/data_tab/secondary_panel.tsx:195]: The toast error pattern is not ideal in the first place because even though the error persists, the toast will disappear. Also this is an error boundary that the user should no longer run into. Its just a safety mechanism to make sure that if it ever does, the whole app does not crash, since we inherit the `DefaultAggParams` component from visualize. 

Also checking with @KrooshalUX about UX signoff.

File: src/plugins/vis_builder/public/types.ts (hunk 2 above)
Comment (ashwin-pc) [reply src/plugins/vis_builder/public/types.ts:40]: Ah this is my bad. For the `filters` aggregation to work, it relies on the app name and storage option in services. Since app name was already being referenced in other places, I decided to add it to the root level service and simply reference that everywhere else too. 

The filters agg uses storage (here it is local storage) with the app name as key to save and retrieve the history of all queries.

File: src/plugins/vis_type_table/public/index.ts (hunk 3 above)
Comment (ashwin-pc) [reply src/plugins/vis_type_table/public/index.ts:14]: Sorry I slipped this in because there is a TS error related to this in the table agg fn which references it. This fixes that TS issue

**Review by joshuarrrr — APPROVED** [review]
Fixing this bug for 2.5.0 seems worthwhile. But can we open a follow-up issue for adding test coverage for the error boundary?

**Review by abbyhu2000 — APPROVED** [review]

File: src/plugins/vis_builder/public/application/components/data_tab/secondary_panel.tsx (hunk 1 above)
Comment (ashwin-pc) [reply src/plugins/vis_builder/public/application/components/data_tab/secondary_panel.tsx:195]: @KrooshalUX is okay with this UX but would also like a toast message to be shown as a part of the error. However that is not a blocker for this and I've created an [issue](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3231) to discuss the use of toasts in error boundaries.

**Comment by ashwin-pc** [comment]
> Fixing this bug for 2.5.0 seems worthwhile. But can we open a follow-up issue for adding test coverage for the error boundary?

Given that this is a standard error boundary, what you want tested here?

File: src/plugins/vis_type_table/public/index.ts (hunk 3 above)
Comment (ananzh) [reply src/plugins/vis_type_table/public/index.ts:14]: Thank you!
